miniconf/src/key.rs

```diff
@@ -54,33 +54,57 @@ impl Key for &str {
     }
 }
 
-/// Capability to yield [`Key`]s
+/// Capability to yield and look up [`Key`]s
 pub trait Keys {
     /// Look up the next key in a [`KeyLookup`] and convert to `usize` index.
     fn next<M: KeyLookup + ?Sized>(&mut self) -> Result<usize, Traversal>;
 
-    /// Return whether there are more keys.
-    ///
-    /// This may mutate and consume remaining keys.
-    fn is_empty(&mut self) -> bool;
+    /// Finalize the keys, ensure there are no more.
+    fn finalize(&mut self) -> bool;
+
+    /// Chain another `Keys` to this one.
+    fn chain<U: IntoKeys>(self, other: U) -> Chain<Self, U::IntoKeys>
+    where
+        Self: Sized,
+    {
+        Chain(self, other.into_keys())
+    }
 }
 
-impl<T> Keys for T
+/// [`Keys`]/[`IntoKeys`] for Iterators of [`Key`]
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+#[repr(transparent)]
+pub struct KeysIter<T: ?Sized>(T);
+
+impl<T> Keys for KeysIter<T>
 where
-    T: Iterator,
+    T: Iterator + ?Sized,
     T::Item: Key,
 {
     fn next<M: KeyLookup + ?Sized>(&mut self) -> Result<usize, Traversal> {
-        let key = Iterator::next(self).ok_or(Traversal::TooShort(0))?;
+        let key = self.0.next().ok_or(Traversal::TooShort(0))?;
         key.find::<M>().ok_or(Traversal::NotFound(1))
     }
 
-    fn is_empty(&mut self) -> bool {
-        self.next().is_none()
+    fn finalize(&mut self) -> bool {
+        self.0.next().is_none()
+    }
+}
+
+impl<T> Keys for &mut T
+where
+    T: Keys + ?Sized,
+{
+    fn next<M: KeyLookup + ?Sized>(&mut self) -> Result<usize, Traversal> {
+        T::next::<M>(self)
+    }
+
+    fn finalize(&mut self) -> bool {
+        T::finalize(self)
     }
 }
 
-/// Capability to be converted into a [`Keys`]
+/// Be converted into a `Keys`
```

Comment:
Why are we using this custom `IntoKeys` trait instead of the bound `core::convert::Into<Keys>`? Is it because `Keys` doesn't have a known type?

Reply:
Rationale is the same as for `IntoIterator`. One reason is indeed to have `Keys` as an associated type and not as an generic type.